File: src/main/java/mnm/mods/tabbychat/client/extra/spell/SpellingFormatter.java
package mnm.mods.tabbychat.client.extra.spell;

import java.util.Iterator;
import java.util.function.Function;

import com.swabunga.spell.event.SpellCheckEvent;

import mnm.mods.tabbychat.util.Color;
import mnm.mods.tabbychat.util.text.ITextBuilder;
import mnm.mods.tabbychat.util.text.TextBuilder;
import net.minecraft.util.text.ITextComponent;

public class SpellingFormatter implements Function<String, ITextComponent> {

    private final Iterator<SpellCheckEvent> spelling;

    private SpellCheckEvent event;
    private int totalLength;

    public SpellingFormatter(Spellcheck sp) {
        spelling = sp.getErrors().iterator();
    }

    @Override
    public ITextComponent apply(String text) {
        if (text == null)
            return null;
        ITextBuilder b = new TextBuilder();
        int prev = 0;
        int length = totalLength;
        // save where we are at.
        totalLength += text.length();
        while (spelling.hasNext() || event != null) {
            if (event == null)
                event = spelling.next();
            int start = event.getWordContextPosition() - length;
            int end = start + event.getInvalidWord().length();

            if (start < 0) {
                // error started on previous line
                start = prev;
            }
            if (start > text.length()) {
                // no more errors on this line
                break;
            }
            b.text(text.substring(prev, start));

            if (end > text.length()) {
                // error goes to next line
                return b.text(text.substring(start)).underline(Color.RED).build();
            }
            b.text(text.substring(start, end)).underline(Color.RED);

            prev = end;
            event = null;
        }
        // no more errors.
        totalLength++;
        return b.text(text.substring(prev)).build();
    }

}

File: src/main/kotlin/mnm/mods/tabbychat/client/extra/spell/SpellingFormatter.kt
package mnm.mods.tabbychat.client.extra.spell

import com.swabunga.spell.event.SpellCheckEvent
import mnm.mods.tabbychat.client.TabbyChatClient
import mnm.mods.tabbychat.util.Color
import mnm.mods.tabbychat.util.text.FancyText
import mnm.mods.tabbychat.util.toComponent
import net.minecraft.util.text.ITextComponent
import net.minecraft.util.text.StringTextComponent

class SpellingFormatter(sp: Spellcheck) {

    private val spelling: Iterator<SpellCheckEvent> = sp.getErrors().iterator()

    private var event: SpellCheckEvent? = null
    private var totalLength: Int = 0

    fun apply(text: String?): ITextComponent? {
        if (text == null)
            return null
        if (text.contains("\u00a7") || !TabbyChatClient.settings.advanced.spelling) {
            return StringTextComponent(text)
        }

        val b = StringTextComponent("")
        var prev = 0
        val length = totalLength
        // save where we are at.
        totalLength += text.length
        while (spelling.hasNext() || event != null) {
            if (event == null)
                event = spelling.next()
            var start = event!!.wordContextPosition - length
            val end = start + event!!.invalidWord.length

            if (start < 0) {
                // error started on previous line
                start = prev
            }
            if (start > text.length) {
                // no more errors on this line
                break
            }
            b.appendText(text.substring(prev, start))

            if (end > text.length) {
                // error goes to next line
                return b.appendSibling(FancyText(text.substring(start).toComponent()).fancyStyle {
                    underline = Color.RED
                })
            }
            b.appendSibling(FancyText(text.substring(start, end).toComponent()).fancyStyle {
                underline = Color.RED

                prev = end
                event = null
            })
        }
        // no more errors.
        totalLength++
        return b.appendText(text.substring(prev))
    }
}